[Site Isolation] _doAfterNextPresentationUpdate does not invoke updateBlock

szewai opened the pull request:
#### 8cbd42ddce32d874cb5ecf8450ea858c23ddd57f
<pre>
[Site Isolation] _doAfterNextPresentationUpdate does not invoke updateBlock
<a href="https://rdar.apple.com/138875168">rdar://138875168</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=294312">https://bugs.webkit.org/show_bug.cgi?id=294312</a>

Reviewed by Alex Christensen.

When site isolation is enabled, the block passed to _doAfterNextPresentationUpdate may not be invoked properly, see the
new API test for an example. There are two causes:
1. When WebPageProxy updates DrawingArea, it does not update DrawingArea of RemotePageProxy, and UI process does not
notify web processes of RemotePageProxy about the update. Therefore, when UI process sends
DrawingArea::DispatchAfterEnsuringDrawing message to those processes, the message is dropped since the web processes are
still tracking old DrawingArea identifier. To fix this, this patch updates DrawingArea of RemotePageProxy in
WebPageProxy::setDrawingArea.
2. In WebPageProxy::callAfterNextPresentationUpdate, UI process sends messages to all web processes that have
RemotePageProxy and wait for them to reply. However, some of them may not reply if their rendering is being suspended
(i.e. RemoteLayerTreeDrawingArea::m_isRenderingSuspended is true), which can happen if the process does not contain an
active frame. To fix this, this patch updates the function to only send to web processes that host frame of the page.

API test: SiteIsolation.DoAfterNextPresentationUpdate

* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::initializeWebPage):
* Source/WebKit/UIProcess/RemotePageDrawingAreaProxy.h:
(WebKit::RemotePageDrawingAreaProxy::identifier const):
* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::setDrawingArea):
* Source/WebKit/UIProcess/RemotePageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::swapToProvisionalPage):
(WebKit::WebPageProxy::setDrawingArea):
(WebKit::WebPageProxy::webContentProcessesWithFrame):
(WebKit::WebPageProxy::callAfterNextPresentationUpdate):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::(SiteIsolation, DoAfterNextPresentationUpdate)):

Canonical link: <a href="https://commits.webkit.org/296313@main">https://commits.webkit.org/296313@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/849ee1b00de0b754f9dbaec5e88aa529e467548c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108118 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27781 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18201 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113329 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58615 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110081 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28477 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36331 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82090 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111066 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22564 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97400 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62522 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/21981 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15536 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58073 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/91925 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15596 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116454 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35183 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/25907 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91115 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35558 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93677 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90910 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23172 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35800 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13563 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/30997 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35084 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40639 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34818 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38176 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36485 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -7287,6 +7292,12 @@ void WebPageProxy::didFinishDocumentLoadForFrame(IPC::Connection& connection, Fr
     }
 }
 
+void WebPageProxy::forEachWebContentProcessWithFrame(NOESCAPE Function<void(WebProcessProxy&)>&& function)
```
As mentioned in the commit message, `forEachWebContentProcess` iterates all process of `RemotePageProxy`, but those processes do not necessarily have active frame of page, and if they don't the rendering of `RemoteLayerTreeDrawingArea` will be suspended. So even those processes have corresponding `DrawingArea`, they are not going to reply `DispatchAfterEnsuringDrawing` since they are not drawing.

It's a good point that this will lead to sending message to the same process multiple times if it contains multiple frames, will do some de-duplication here.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/16d0334cf1c97335b2500b40e783c77caeb56491. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46592

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c95fa7052c8077f22125c048df02badb1cb168ac. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46592

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/05a9b2ca3fabeec16d72d1541ad28b0db19bb883. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46592

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/849ee1b00de0b754f9dbaec5e88aa529e467548c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46592

szewai committed:
[Site Isolation] _doAfterNextPresentationUpdate does not invoke updateBlock
rdar://138875168
https://bugs.webkit.org/show_bug.cgi?id=294312

Reviewed by Alex Christensen.

When site isolation is enabled, the block passed to _doAfterNextPresentationUpdate may not be invoked properly, see the
new API test for an example. There are two causes:
1. When WebPageProxy updates DrawingArea, it does not update DrawingArea of RemotePageProxy, and UI process does not
notify web processes of RemotePageProxy about the update. Therefore, when UI process sends
DrawingArea::DispatchAfterEnsuringDrawing message to those processes, the message is dropped since the web processes are
still tracking old DrawingArea identifier. To fix this, this patch updates DrawingArea of RemotePageProxy in
WebPageProxy::setDrawingArea.
2. In WebPageProxy::callAfterNextPresentationUpdate, UI process sends messages to all web processes that have
RemotePageProxy and wait for them to reply. However, some of them may not reply if their rendering is being suspended
(i.e. RemoteLayerTreeDrawingArea::m_isRenderingSuspended is true), which can happen if the process does not contain an
active frame. To fix this, this patch updates the function to only send to web processes that host frame of the page.

API test: SiteIsolation.DoAfterNextPresentationUpdate

* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::initializeWebPage):
* Source/WebKit/UIProcess/RemotePageDrawingAreaProxy.h:
(WebKit::RemotePageDrawingAreaProxy::identifier const):
* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::setDrawingArea):
* Source/WebKit/UIProcess/RemotePageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::swapToProvisionalPage):
(WebKit::WebPageProxy::setDrawingArea):
(WebKit::WebPageProxy::webContentProcessesWithFrame):
(WebKit::WebPageProxy::callAfterNextPresentationUpdate):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::(SiteIsolation, DoAfterNextPresentationUpdate)):

Canonical link: https://commits.webkit.org/296313@main

webkit-commit-queue commented:
Committed 296313@main (8cbd42ddce32): <https://commits.webkit.org/296313@main>

Reviewed commits have been landed. Closing PR #46592 and removing active labels.
